pkg/recordbot/recorder.go

```diff
@@ -0,0 +1,136 @@
+package recordbot
+
+import (
+	"errors"
+	"log"
+
+	"github.com/pion/rtp"
+	"github.com/pion/webrtc/v3"
+	"github.com/pion/webrtc/v3/pkg/media"
+	"github.com/pion/webrtc/v3/pkg/media/h264writer"
+	"github.com/pion/webrtc/v3/pkg/media/ivfwriter"
+	"github.com/pion/webrtc/v3/pkg/media/oggwriter"
+)
+
+var ErrUnsupportedCodec = errors.New("unsupported codec")
+
+type Recorder interface {
+	// Start reads RTP packets in TrackRemote and save them to local file.
+	// No need to call this in a goroutine as the function already calls one.
+	Start(track *webrtc.TrackRemote)
+
+	// Stop the recording and does cleanups such as closing the file and channel.
+	Stop()
+}
+
+type recorder struct {
+	// Save file name to be passed into hooks
+	filename string
+
+	// Each recorder uses different writer depending if it's for video / audio
+	writer media.Writer
+
+	// Channels for tracking the state of the recorder
+	done   chan struct{}
+	closed chan struct{}
+
+	// Hooks provide access to customising aspects of the recorder, such as uploading the file after recording
+	hooks RecorderHooks
+}
+
+type RecorderHooks struct {
+	// The recorder will pass the filename that it was instantiated with, to the hook
+	UploadFile func(filename string) error
+}
+
+func NewSingleTrackRecorder(filename string, codec webrtc.RTPCodecParameters, hooks RecorderHooks) (Recorder, error) {
+	// Instantiates the appropriate media writer to save RTP packets
+	writer, err := createMediaWriter(filename, codec.MimeType)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make channels for tracking recorder state
+	done := make(chan struct{}, 1)
+	closed := make(chan struct{}, 1)
+
+	return &recorder{filename, writer, done, closed, hooks}, nil
+}
+
+func createMediaWriter(fileName string, mimeType string) (media.Writer, error) {
+	if mimeType == webrtc.MimeTypeVP8 || mimeType == webrtc.MimeTypeVP9 {
+		return ivfwriter.New(fileName)
+	}
+	if mimeType == webrtc.MimeTypeH264 {
+		return h264writer.New(fileName)
+	}
+	if mimeType == webrtc.MimeTypeG722 ||
+		mimeType == webrtc.MimeTypePCMA ||
+		mimeType == webrtc.MimeTypePCMU ||
+		mimeType == webrtc.MimeTypeOpus {
+		// The values for oggwriter can be found at
+		// https://github.com/pion/webrtc/blob/master/examples/save-to-disk/main.go#L53
+		return oggwriter.New(fileName, 48000, 0)
```

Comment:
channelCount should be either 1 or 2, or `.ogg` file gets created with muted audio. The linked example uses OggWriter with 2 channels.

Reply:
Hi @rrampage , thanks for the catch! I wouldn't know about it.